app/services/qa/pagination_service.rb

```diff
@@ -0,0 +1,513 @@
+      #       startRecord (deprecated, supported for backward compatibility with
+      #       linked_data module pagination).
+      def requested_page_offset
+        return @requested_page_offset if @requested_page_offset.present?
+        @requested_page_offset = (request_params['page_offset'] || request_params['startRecord'])
+      end
+
+      # @return [Integer] The first record to include in the response data as
+      #     requested as long as it is an Integer; otherwise, returns nil.
+      def validated_request_page_offset
+        @page_offset_error = false
+        offset = Integer(requested_page_offset)
+        return offset if offset == 1
+        @page_offset_error = ERROR_OUT_OF_RANGE_TOO_SMALL if offset < 1
+        @page_offset_error = ERROR_OUT_OF_RANGE_TOO_LARGE if offset > total_num_found
+        offset
+      rescue
+        @page_offset_error = ERROR_NOT_INTEGER
+        nil
+      end
+
+      # @return [Integer] The requested maximum number of results to return (default=DEFAULT_PAGE_LIMIT | ALL)
+      # @see #default_page_limit
+      # @see DEFAULT_PAGE_LIMIT
+      def page_limit
+        return @page_limit if @page_limit.present?
+        return @page_limit = default_page_limit unless page_limit_specified?
+        @page_limit = validated_request_page_limit || default_page_limit
+      end
+
+      # @return [Boolean] true if the request specifies the page limit; otherwise, false
+      def page_limit_specified?
+        request_params.keys.include?("page_limit") || request_params.keys.include?("maxRecords")
+      end
+
+      # @return [Integer] The max number of records for a page as specified in the request_params.
+      # @note The page size limit can be specified with page_limit (preferred) or
+      #       maxRecords (deprecated, supported for backward compatibility with
+      #       linked_data module pagination).
+      def requested_page_limit
+        return @requested_page_limit if @requested_page_limit.present?
+        @requested_page_limit ||= (request_params['page_limit'] || request_params['maxRecords'])
+      end
+
+      # @return [Integer] The max number of records to include in response data as
+      #     requested as long as it is a positive Integer; otherwise, returns nil.
+      def validated_request_page_limit
+        @page_limit_error = false
+        limit = Integer(requested_page_limit)
+        @page_limit_error = ERROR_OUT_OF_RANGE_TOO_SMALL if limit < 1
+        limit.positive? ? limit : nil
+      rescue
+        @page_limit_error = ERROR_NOT_INTEGER
+        nil
+      end
+
+      # @return [Integer] The default to use when page_limit isn't specified.
+      # @note To maintain backward compatibility, the limit will be all results
+      #   if format is json and neither page_limit nor page_offset were specified.
+      def default_page_limit
+        return total_num_found unless json_api? || page_offset_specified? || page_limit_specified?
```

Comment:
I started off having `page_limit=total_num_found` if `format==:json` and `page_limit` is not passed in regardless of whether `page_offset` was passed in.  As I thought about it more, it seemed to me that if `page_offset` is passed in, then it implies a desire for pagination, in which case, the default `page_limit` should be used.

Reply:
I think that totally makes sense and provides the same pagination behavior across both the `:json` and `:json_api` formats.

Reply:
Here is an example of how it would look with 36 results if the default for `:json` was to default `page_limit=total_num_found`…

- page_offset=1 - returns all [1..36]
- page_offset=11 - returns [11..36]
- page_offset=21 - returns [21..36]
- page_offset=31 - returns [31-36]

As opposed to using the default `page_limit` (i.e., 10)…

- page_offset=1 - returns [1..10]
- page_offset=11 - returns [11..20]
- page_offset=21 - returns [21..30]
- page_offset=31 - returns [31-36]